File: simplify-path/solution71.go
// https://leetcode.com/problems/simplify-path/

package main

import (
	"fmt"
	"strings"
)

// T: O(n), S: O(n)
func simplifyPath(path string) string {
	stack := make([]string, 0)
	paths := strings.Split(path, "/")
	for _, p := range paths {
		if len(p) == 0 || p == "." {
			continue
		}
		if p == ".." {
			if len(stack) != 0 {
				stack = stack[:len(stack)-1]
			}
		} else {
			stack = append(stack, p)
		}
	}
	return "/" + strings.Join(stack, "/")
}

func main() {
	path1 := "/home/"
	path2 := "/../"
	path3 := "/home//foo/"
	fmt.Println(simplifyPath(path1)) // "/home"
	fmt.Println(simplifyPath(path2)) // "/"
	fmt.Println(simplifyPath(path3)) // "/home/foo"
}

File: simplify-path/Solution71.java
// https://leetcode.com/problems/simplify-path/

import java.util.Stack;

class Solution71 {

    // T: O(n), S: O(n)
    public String simplifyPath(String path) {
        var stack = new Stack<String>();
        var paths = path.split("/");
        for (var p : paths) {
            if (p.isEmpty() || p.equals(".")) {
                continue;
            }
            if (p.equals("..")) {
                if (!stack.isEmpty()) {
                    stack.pop();
                }
            } else {
                stack.push(p);
            }
        }
        var sb = new StringBuilder();
        for (var p : stack) {
            sb.append("/");
            sb.append(p);
        }
        return sb.isEmpty() ? "/" : sb.toString();
    }

    public static void main(String[] args) {
        var sol = new Solution71();
        var path1 = "/home/";
        var path2 = "/../";
        var path3 = "/home//foo/";
        System.out.println(sol.simplifyPath(path1)); // "/home"
        System.out.println(sol.simplifyPath(path2)); // "/"
        System.out.println(sol.simplifyPath(path3)); // "/home/foo"
    }

}
